pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/EtcdMetadataStore.java

```diff
@@ -0,0 +1,423 @@
+                                                if (log.isDebugEnabled()) {
+                                                    log.debug("On next: {}", leaseKeepAliveResponse);
+                                                }
+                                            }
+
+                                            @Override
+                                            public void onError(Throwable throwable) {
+                                                log.warn("Lease client error :", throwable);
+                                                receivedSessionEvent(SessionEvent.SessionLost);
+                                            }
+
+                                            @Override
+                                            public void onCompleted() {
+                                                log.info("Etcd lease has expired");
+                                                receivedSessionEvent(SessionEvent.SessionLost);
+                                            }
+                                        });
+                    }
+                });
+
+        if (retryOnFailure) {
+            future.exceptionally(ex -> {
+                log.warn("Failed to create Etcd lease. Retrying later", ex);
+                executor.schedule(() -> {
+                    createLease(true);
+                }, 1, TimeUnit.SECONDS);
+                return null;
+            });
+        }
+
+        return future;
+    }
+
+    private void handleWatchResponse(WatchResponse watchResponse) {
+        watchResponse.getEvents().forEach(we -> {
+            String path = we.getKeyValue().getKey().toString(StandardCharsets.UTF_8);
+            String parent = parent(path);
+            if (we.getEventType() == WatchEvent.EventType.PUT) {
+                if (we.getKeyValue().getVersion() == 1) {
+                    receivedNotification(new Notification(NotificationType.Created, path));
+
+                    if (parent != null) {
+                        receivedNotification(new Notification(NotificationType.ChildrenChanged, parent));
+                    }
+                } else {
+                    receivedNotification(new Notification(NotificationType.Modified, path));
+                }
+            } else if (we.getEventType() == WatchEvent.EventType.DELETE) {
+                receivedNotification(new Notification(NotificationType.Deleted, path));
+                if (parent != null) {
+                    receivedNotification(new Notification(NotificationType.ChildrenChanged, parent));
+                }
+            }
+        });
+    }
+
+    @Override
+    protected void receivedSessionEvent(SessionEvent event) {
+        if (event == SessionEvent.SessionReestablished) {
+            // Re-create the lease before notifying that we are reconnected
+            createLease(true)
+                    .thenRun(() -> {
+                        super.receivedSessionEvent(event);
```

Comment:
If `createLease` failed the first time, the `receivedSessionEvent` will not be called.

Reply:
Yes, that is intended. 

The code here can certainly be refactored to make it easier. The intention here is: 
 * At startup, we create the lease. If we fail, we bubble up the exception (eg: broker fails to start)
 * If we lose the lease while running, depending on the configuration we either kill the broker or try to reconnect.
 * When we don't kill the broker, we re-establish a new lease and we keep trying doing that indefinitely

Reply:
> * When we don't kill the broker, we re-establish a new lease and we keep trying doing that indefinitely

My point is, in this case,  if we finally re-establish the new lease (after some retries), is it ok that we don't call `super.receivedSessionEvent(event)`?  Current CompletableFuture returned by `createLease ` only tracks the first try.

Reply:
After reconnection, the `receivedSessionEvent()` will be called through the session watcher class.

Reply:
I see, thx~